### PR TITLE
Add Cardano chain

### PR DESCRIPTION
**Pull request by FilippoScaramuzza**
**Description**: added Cardano chain.

**Ticket(s)**: #1 (closed but should be the actual one) and #16.

**Contribution Details**:
I used the [Cexplorer APIs](https://cexplorer.io/devs) to get the list of pools and staked amount.

From Cexplorer, the Nakamoto coefficient is [something like 36](https://cexplorer.io/nakamoto). In the PR, anyway, the value you will get is expected to be higher (the last time I tried was 112) as the single pools are used instead of the [groups](https://cexplorer.io/groups); that is what they used (confirmed by an admin on Discord).

### CODE REVIEW

**Comment by xenowits** [comment]
Hey @FilippoScaramuzza , thanks for submitting this PR!

The NC value comes out to be 112, while the cexplorer seems to report 36. These data look very different, so I'd say we should look into how the cexplorer guys calculate it.

Would this be something you could take up?

**Comment by FilippoScaramuzza** [comment]
Hi!
I asked exactly this to the admins. This is what they said:
image

They're using groups of pools instead of single ones. I guess it's a different point of view, nonetheless still appropriate. 
I like to focus on the single pools instead of referring to the groups. Also, group statistics need to be wholly automated and periodically manually updated. Hence, it might not be a reliable source of information.

More information about groups here: [https://cexplorer.io/link/donuts](https://cexplorer.io/link/donuts).

**Comment by alfredonodo** [comment]
Hello,
I used the data provided [here](https://cexplorer.io/groups). There are 178 staking pool operators, adding up the total stake of each staking pool and removing the single pool operators, the actual Nakamoto coefficient is 35. So the number shown [here](https://cexplorer.io/nakamoto) is correct.
In Cardano, a [staking pool operator](https://cardano.org/stake-pool-operation/) is a single entity and acts as a validator for other networks. The consensus protocol requires [50% of the stake](https://arxiv.org/abs/1904.04098).

**Comment by Straightpool** [comment]
> Hello, I used the data provided [here](https://cexplorer.io/groups). There are 178 staking pool operators, adding up the total stake of each staking pool and removing the single pool operators, the actual Nakamoto coefficient is 35. 

The data source you use is good and often referenced in the Cardano community, however there are much more than 178 staking pool operators in Cardano. The largest group "Single Pool Operators" bundles up all single staking pools aka pools where the operator is believed to only operate one single pool. The other segments show so called "Multi Pool Operators" so a single operator or an organisation operating more than one pool, usually under the same brand or same marketing, a cluster of pools. How many pools are hidden behind each donut chart segment is shown in the tooltip, so if you hover on "Single Pool Operators", you can see there are currently 1815 operators in this group. 
For the calculation of the MAV obviously the big clusters are counted or multi pool operators counting from the largest cluster until 50% of stake is reached.

Here at https://pooltool.io/ in the top navigation you can see the total number of pool with at least one LoveLace staked, currently 2871 pools. One ADA is 1M Lovelace.

**Comment by alfredonodo** [comment]
> > Hello, I used the data provided [here](https://cexplorer.io/groups). There are 178 staking pool operators, adding up the total stake of each staking pool and removing the single pool operators, the actual Nakamoto coefficient is 35.
> 
> The data source you use is good and often referenced in the Cardano community, however there are much more than 178 staking pool operators in Cardano. The largest group "Single Pool Operators" bundles up all single staking pools aka pools where the operator is believed to only operate one single pool. The other segments show so called "Multi Pool Operators" so a single operator or an organisation operating more than one pool, usually under the same brand or same marketing, a cluster of pools. How many pools are hidden behind each donut chart segment is shown in the tooltip, so if you hover on "Single Pool Operators", you can see there are currently 1815 operators in this group. For the calculation of the MAV obviously the big clusters are counted or multi pool operators counting from the largest cluster until 50% of stake is reached.
> 
> Here at https://pooltool.io/ in the top navigation you can see the total number of pools with at least one LoveLace staked, currently 2871 pools. One ADA is 1M Lovelace.

Thank you very much for the detailed explanation. As you know, what is important to determine the PoS consensus and thus the generation and validation of blocks is not the total number of nodes, but the minimum number that decides i.e. the so-called Nakamoto coefficient. For example, on Ethereum there are about 900k nodes, but only about 10 of them rule the network.

**Comment by thenic95** [comment]
Hi @xenowits ,
Hope all is well!
How can I help to get this PR ready for merge? 
Thanks for your help

**Comment by xenowits** [comment]
Yes, sure @thenic95 !

Please update the code as pointed by @alfredonodo to also include staking pool operators. I can review from there

**Comment by MokhFn** [comment]
Is this still ongoing ? Can we move forward with this, happy to help.

**Comment by thenic95** [comment]
I was not able to update the code @MokhFn. Would you be able to help?

File: core/chains/cardano.go
Comment (FilippoScaramuzza) [reply core/chains/cardano.go:36]: Maybe we can add it as a parameter:

```shell
docker run --rm \
-e "SOLANA_API_KEY=<YOUR_SOLANA_API_KEY_HERE>" \
-e "CARDANO_API_KEY=<YOUR_CARDANO_API_KEY_HERE>" \
-e "RATED_API_KEY=<YOUR_RATED_API_KEY_HERE>" \
-p 8080:8080 xenowits/nc-calc:v0.1.4
```

Anyway, I'm not completely sure about the validity.
While searching for another data source to close this PR, I came across [another PR](https://github.com/xenowits/nakomoto-coefficient-calculator/issues/47). Here, [@ccgarant](https://github.com/ccgarant), suggested to use https://www.balanceanalytics.io/chartboards/donut_shop. I searched the data source in the networking tab and noticed the API call to https://api.balanceanalytics.io/rpc/pool_group_stake_donut, where the bearer token to get authenticated can be found in the page's source code as `PUBLIC_BALANCE_API_TOKEN`:
```javascript
__sveltekit_12bou38 = {
  base: new URL("..",location).pathname.slice(0, -1),
  env: {
      "PUBLIC_BALANCE_API_TOKEN": "...",
      "PUBLIC_SUPABASE_ANON_KEY": "...",
  ...
  }
};
```
So they're apparently loading it from the environment and then using it. Anyway, I'm not really sure if this is intentional or not. I'll try to contact @ccgarant and see if he can help.

**Comment by FilippoScaramuzza** [comment]
I've got some time to work on this. 
After the discussion with @alfredonodo in this thread and after reading the thread in PR #47, I searched for another data source taking into account pool groups. 
The website mentioned in #47 is pretty well done [https://www.balanceanalytics.io/chartboards/donut_shop](https://www.balanceanalytics.io/chartboards/donut_shop) and there is an API [https://api.balanceanalytics.io/rpc/pool_group_stake_donut](https://api.balanceanalytics.io/rpc/pool_group_stake_donut) that returns the list of groups and associated stakes. (I'm not so sure if we are actually allowed to use that API as it is not listed and requires a token that I had to read directly from the page's source code, see [comment above](https://github.com/xenowits/nakomoto-coefficient-calculator/pull/43#discussion_r1662656299)).

Anyway, as mentioned by @alfredonodo, the threshold for Cardano is not 0.33, but 0.50.  
Therefore, I created another function
```
func CalcNakamotoCoefficientBigNums51(totalVotingPower *big.Int, votingPowers []big.Int) int {
	thresholdPercent := big.NewFloat(0.51)
	thresholdVal := new(big.Float).Mul(new(big.Float).SetInt(totalVotingPower), thresholdPercent)
	cumulativeVal := big.NewFloat(0.00)
	nakamotoCoefficient := 0

	for _, vp := range votingPowers {
		z := new(big.Float).Add(cumulativeVal, new(big.Float).SetInt(&vp))
		cumulativeVal = z
		nakamotoCoefficient += 1
		if cumulativeVal.Cmp(thresholdVal) == +1 {
			break
		}
	}

	return nakamotoCoefficient
}
```
where the logic is exactly the same as `CalcNakamotoCoefficientBigNums`. The idea (I think we can do it in another PR) is to have a parameter called `threshold` that we can adjust and modify based on the specific network we are dealing with to make it more elegant and scalable.

**Comment by GilbertStClr** [comment]
Hello and thank you for your patience here.  I have created the API endpoint per your requirements.  It can be called using the following URL: https://www.balanceanalytics.io/api/mavdata.json

**Comment by GilbertStClr** [comment]
Please do test the data provided via the API mentioned above.  If any changes are required, please let me know and I will be sure to update.  The API data will be updated within the 24 hours following the start of a new epoch.

**Comment by FilippoScaramuzza** [comment]
Thank you @GilbertStClr! The API works perfectly, I just pushed the updated code! 😄 

I noticed a typo in the new function `CalcNakamotoCoefficientBigNums51` where the threshold was mistakenly set to 0.51 instead of 0.50. However, I believe that the name with the 51 is still appropriate.

Let me know if other changes are needed. Thanks!!

**Comment by MokhFn** [comment]
@FilippoScaramuzza I'd update [this doc](https://github.com/xenowits/nakomoto-coefficient-calculator/blob/main/README.md) by adding the [cardano foundation's](https://cardano.org) website to the list of supported chains. Other than that LGTM.
Nice work ! It's left to the maintainers to approve i guess.

**Comment by FilippoScaramuzza** [comment]
Nice, thank you. I'll wait for @xenowits for the final review and how to update [the doc](https://github.com/xenowits/nakomoto-coefficient-calculator/blob/main/README.md) (Cardano has a different threshold, so we should change or generalize the README).

**Comment by GilbertStClr** [comment]
Just as an FYI, I have updated the endpoint to include a new field called "class" which identifies the pool as MPO or sSPO per BALANCE pool group data set.

**Review by MokhFn — CHANGES_REQUESTED** [review]
Doc update missing : [this doc](https://github.com/xenowits/nakomoto-coefficient-calculator/blob/main/README.md) by adding the [cardano foundation's](https://cardano.org/) website to the list of supported chains. Other than that LGTM.
To confirm with @xenowits for the threshold.

**Comment by FilippoScaramuzza** [comment]
I've updated the code by modifying the data structure to reflect the changes outlined by @GilbertStClr. I modified the [README.md](https://github.com/xenowits/nakomoto-coefficient-calculator/blob/main/README.md) including Cardano and its website. Let me know if any other changes are needed.

**Comment by FilippoScaramuzza** [comment]
Hi!
Any updates on this @xenowits?

**Comment by alfredonodo** [comment]
Hi,
any updates on this @xenowits @FilippoScaramuzza ?
Thank you

**Comment by ccgarant** [comment]
Wen.

**Comment by hodlonaut** [comment]
Any progress updates on this PR, folks? What seems to be the hold-up? ccing @xenowits

**Comment by xenowits** [comment]
Just checking this out. Apologies for the delay here.

**Comment by xenowits** [comment]
@FilippoScaramuzza Let's work to merge this long-standing PR. Can you please:
1. Pull latest changes and fix merge conflicts?
2. After that, could you please run the command share the screenshot for the NC value for Cardano?

**Comment by FilippoScaramuzza** [comment]
Hi @xenowits, thanks for the update.
Working on it, I'll push asap!

**Comment by FilippoScaramuzza** [comment]
Hi @xenowits ,
merge conflicts should be fine. 
Here's the screenshot:
image

Let me know if you need anything else!

**Comment by xenowits** [comment]
@FilippoScaramuzza LGTM! If you don't have anything else you want to fix, I'll go ahead and merge this.

File: README.md
Comment (xenowits) [inline README.md:66]: @FilippoScaramuzza this is an alphabetical ordering. pls move this accordingly.

File: core/utils/calc_nakamoto_coefficient_big_nums.go
Comment (xenowits) [inline core/utils/calc_nakamoto_coefficient_big_nums.go:29]: pls add a comment for this function

**Review by xenowits — APPROVED** [review]

File: README.md
Comment (FilippoScaramuzza) [reply README.md]: I changed the chains order, so now it's alphabetical.

**Comment by FilippoScaramuzza** [comment]
I don't have any other specific things to fix (just fixed with your last comments).
You can go ahead!
Thanks.

**Comment by ccgarant** [comment]
The site appears down?

**Comment by xenowits** [comment]
@ccgarant The site going down has nothing to do with this merge. This change will be deployed in the new year, after the holidays. The deployment currently is frozen.

**Pull request merged**
